File: tools/invisible_unicode_lint/src/lint-invisible-unicode.ts
/**
 * Deterministic scanner for invisible / smuggling Unicode in text ingested by humans and agents.
 * No network, no models: codepoint checks from patterns JSON.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { parseArgs } from "node:util";

export interface Rule {
  readonly ruleId: string;
  readonly start: number;
  readonly end: number;
  readonly title: string;
  readonly detail: string;
  readonly allowOnlyAtFileStart: boolean;
}

export interface PatternsFile {
  version?: number;
  extensions?: string[];
  exclude_path_substrings?: string[];
  rules?: Array<{
    id: string;
    start: number;
    end: number;
    title: string;
    detail: string;
    allow_only_at_file_start?: boolean;
  }>;
}

export function loadPatterns(patternsPath: string): {
  rules: Rule[];
  extensions: Set<string>;
  excludeSubstrings: string[];
} {
  const raw = JSON.parse(fs.readFileSync(patternsPath, "utf8")) as PatternsFile;
  const rules: Rule[] = (raw.rules ?? []).map((item) => ({
    ruleId: String(item.id),
    start: Number(item.start),
    end: Number(item.end),
    title: String(item.title),
    detail: String(item.detail),
    allowOnlyAtFileStart: Boolean(item.allow_only_at_file_start),
  }));
  const extensions = new Set(
    (raw.extensions ?? []).map((e) => e.toLowerCase()),
  );
  const excludeSubstrings = raw.exclude_path_substrings ?? [];
  return { rules, extensions, excludeSubstrings };
}

function matchRules(code: number, offset: number, rules: Iterable<Rule>): Rule[] {
  const matched: Rule[] = [];
  for (const r of rules) {
    if (r.start <= code && code <= r.end) {
      if (r.allowOnlyAtFileStart && offset === 0) {
        continue;
      }
      matched.push(r);
    }
  }
  return matched;
}

/** Each violation: line, column, codepoint, rule (line/column 1-based). */
export function scanText(text: string, rules: Iterable<Rule>): Array<[number, number, number, Rule]> {
  const violations: Array<[number, number, number, Rule]> = [];
  let line = 1;
  let col = 1;
  let cpOffset = 0;
  for (const ch of text) {
    const code = ch.codePointAt(0)!;
    for (const rule of matchRules(code, cpOffset, rules)) {
      violations.push([line, col, code, rule]);
    }
    if (ch === "\n") {
      line++;
      col = 1;
    } else if (ch === "\r") {
      /* keep column */
    } else {
      col++;
    }
    cpOffset++;
  }
  return violations;
}

export function formatViolation(
  filePath: string,
  line: number,
  column: number,
  code: number,
  rule: Rule,
): string {
  return `${filePath}:${line}:${column}: ${rule.ruleId} U+${code.toString(16).toUpperCase().padStart(4, "0")} (${rule.title}) — ${rule.detail}`;
}

function shouldScan(
  filePath: string,
  extensions: Set<string>,
  excludeSubstrings: string[],
): boolean {
  const normalized = filePath.split(path.sep).join("/");
  for (const frag of excludeSubstrings) {
    if (normalized.includes(frag)) {
      return false;
    }
  }
  const ext = path.extname(filePath).toLowerCase();
  return extensions.has(ext);
}

function* walkFiles(rootDir: string): Generator<string> {
  const entries = fs.readdirSync(rootDir, { withFileTypes: true });
  for (const ent of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    const full = path.join(rootDir, ent.name);
    if (ent.isDirectory()) {
      yield* walkFiles(full);
    } else if (ent.isFile()) {
      yield full;
    }
  }
}

export function iterFiles(
  roots: string[],
  extensions: Set<string>,
  excludeSubstrings: string[],
): string[] {
  const out = new Set<string>();
  for (const root of roots) {
    const abs = path.resolve(root);
    if (!fs.existsSync(abs)) {
      continue;
    }
    const st = fs.statSync(abs);
    if (st.isFile()) {
      if (shouldScan(abs, extensions, excludeSubstrings)) {
        out.add(abs);
      }
    } else if (st.isDirectory()) {
      for (const f of walkFiles(abs)) {
        if (shouldScan(f, extensions, excludeSubstrings)) {
          out.add(f);
        }
      }
    }
  }
  return [...out].sort();
}

function readUtf8Strict(filePath: string): string | null {
  const buf = fs.readFileSync(filePath);
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(buf);
  } catch {
    return null;
  }
}

export function defaultPatternsPath(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  return path.join(here, "..", "patterns.default.json");
}

export function main(argv: string[]): number {
  const { values, positionals } = parseArgs({
    args: argv,
    options: {
      patterns: { type: "string", short: "p" },
    },
    allowPositionals: true,
  });

  const patternsPath = values.patterns ?? defaultPatternsPath();
  const { rules, extensions, excludeSubstrings } = loadPatterns(patternsPath);
  const roots = positionals.length > 0 ? positionals : ["."];

  const files = iterFiles(roots, extensions, excludeSubstrings);
  let exitCode = 0;

  for (const filePath of files) {
    const text = readUtf8Strict(filePath);
    if (text === null) {
      console.error(`${filePath}:0:0: invalid-utf8 — file is not valid UTF-8`);
      exitCode = 1;
      continue;
    }
    for (const [line, col, code, rule] of scanText(text, rules)) {
      console.error(formatViolation(filePath, line, col, code, rule));
      exitCode = 1;
    }
  }

  return exitCode;
}

function invokedAsCli(): boolean {
  const entry = process.argv[1];
  if (!entry) {
    return false;
  }
  return import.meta.url === pathToFileURL(path.resolve(entry)).href;
}

if (invokedAsCli()) {
  try {
    process.exit(main(process.argv.slice(2)));
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

File: tools/invisible_unicode_lint/lint_invisible_unicode.py
#!/usr/bin/env python3
"""
Deterministic scanner for invisible / smuggling Unicode in text ingested by humans and agents.

No network, no models: pure codepoint checks loaded from patterns JSON.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable, Optional, Sequence


@dataclass(frozen=True)
class Rule:
    rule_id: str
    start: int
    end: int
    title: str
    detail: str
    allow_only_at_file_start: bool


def _load_patterns(path: Path) -> dict[str, Any]:
    with path.open(encoding="utf-8") as f:
        return json.load(f)


def _parse_rules(raw: dict[str, Any]) -> list[Rule]:
    out: list[Rule] = []
    for item in raw.get("rules", []):
        out.append(
            Rule(
                rule_id=str(item["id"]),
                start=int(item["start"]),
                end=int(item["end"]),
                title=str(item["title"]),
                detail=str(item["detail"]),
                allow_only_at_file_start=bool(item.get("allow_only_at_file_start", False)),
            )
        )
    return out


def _match_rules(code: int, offset: int, rules: Iterable[Rule]) -> list[Rule]:
    matched: list[Rule] = []
    for r in rules:
        if r.start <= code <= r.end:
            if r.allow_only_at_file_start and offset == 0:
                continue
            matched.append(r)
    return matched


def scan_text(text: str, rules: Iterable[Rule]) -> list[tuple[int, int, int, Rule]]:
    """
    Returns list of (line, column, codepoint, rule) for each violation.
    Line and column are 1-based.
    """
    violations: list[tuple[int, int, int, Rule]] = []
    line = 1
    col = 1
    for offset, ch in enumerate(text):
        o = ord(ch)
        for rule in _match_rules(o, offset, rules):
            violations.append((line, col, o, rule))
        if ch == "\n":
            line += 1
            col = 1
        elif ch == "\r":
            pass
        else:
            col += 1
    return violations


def _should_scan(path: Path, extensions: set[str], exclude_substrings: tuple[str, ...]) -> bool:
    s = str(path)
    for frag in exclude_substrings:
        if frag in s.replace("\\", "/"):
            return False
    return path.suffix.lower() in extensions


def iter_files(paths: list[Path], extensions: set[str], exclude_substrings: tuple[str, ...]) -> list[Path]:
    files: list[Path] = []
    for p in paths:
        if p.is_file():
            if _should_scan(p, extensions, exclude_substrings):
                files.append(p)
        elif p.is_dir():
            for child in sorted(p.rglob("*")):
                if child.is_file() and _should_scan(child, extensions, exclude_substrings):
                    files.append(child)
    return sorted(set(files))


def format_violation(path: Path, line: int, col: int, code: int, rule: Rule) -> str:
    return (
        f"{path}:{line}:{col}: {rule.rule_id} U+{code:04X} ({rule.title}) — {rule.detail}"
    )


def main(argv: Optional[Sequence[str]] = None) -> int:
    parser = argparse.ArgumentParser(
        description="Fail if files contain high-risk invisible / bidi smuggling Unicode.",
    )
    parser.add_argument(
        "paths",
        nargs="*",
        default=["."],
        help="Files or directories to scan (default: .)",
    )
    parser.add_argument(
        "--patterns",
        type=Path,
        default=Path(__file__).resolve().parent / "patterns.default.json",
        help="JSON rule file (default: packaged patterns.default.json)",
    )
    args = parser.parse_args(argv)

    raw = _load_patterns(args.patterns)
    rules = _parse_rules(raw)
    extensions = {e.lower() for e in raw.get("extensions", [])}
    exclude_substrings = tuple(raw.get("exclude_path_substrings", []))

    paths = [Path(p) for p in args.paths]
    files = iter_files(paths, extensions, exclude_substrings)

    exit_code = 0
    for path in files:
        try:
            text = path.read_text(encoding="utf-8")
        except UnicodeDecodeError as e:
            print(f"{path}:0:0: invalid-utf8 — file is not valid UTF-8 ({e})", file=sys.stderr)
            exit_code = 1
            continue
        for line, col, code, rule in scan_text(text, rules):
            print(format_violation(path, line, col, code, rule), file=sys.stderr)
            exit_code = 1

    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())
